sklearn/gp_search.py

```diff
@@ -0,0 +1,421 @@
+                 n_iter=20,
+                 n_candidates=500,
+                 gp_params={},
+                 verbose=0):
+
+        self.estimator = estimator
+        self.parameters = parameters
+        self.n_parameters = len(parameters)
+        self.scoring = scoring
+        self.acquisition_function = acquisition_function
+        self.n_iter = n_iter
+        self.n_init = n_init
+        self.refit = refit
+        self.n_candidates = n_candidates
+        self.gp_params = gp_params
+        self.param_names = list(parameters.keys())
+
+        # FIXME: 
+        self.param_is_int = np.array([0 if (parameters[k][0] == 'float')
+                                     else 1 for k in self.param_names])
+        self.param_bounds = np.zeros((self.n_parameters, 2))
+        self.verbose = verbose
+        self.fit_params = fit_params if fit_params is not None else {}
+        self.cv = cv
+
+        self.best_params_ = None
+        self.tested_parameters_ = None
+        self.cv_scores_ = None
+
+        if callable(estimator):
+            self._callable_estimator = True
+            if self.verbose > 0:
+                print('Estimator is a callable and not an sklearn Estimator')
+        else:
+            self._callable_estimator = False
+
+        # init param_bounds
+        for i in range(self.n_parameters):
+            if parameters[self.param_names[i]][0] == 'cat':
+                self.param_bounds[i, 0] = 0
+                self.param_bounds[i, 1] = \
+                    len(parameters[self.param_names[i]][1])
+            else:
+                self.param_bounds[i] = \
+                    np.array(parameters[self.param_names[i]][1])
+                if parameters[self.param_names[i]][0] == 'int':
+                    self.param_bounds[i, 1] += 1
+
+        if self.verbose > 0:
+            print(self.parameters)
+            print(self.param_names)
+            print(self.param_is_int)
+            print(self.param_bounds)
+
+    def _vector_to_dict(self, vector_parameter):
+        dict_parameter = dict.fromkeys(self.param_names)
+        for i in range(self.n_parameters):
+            if self.parameters[self.param_names[i]][0] == 'cat':
+                dict_parameter[self.param_names[i]] = \
+                    (self.parameters[self.param_names[i]][1])[
+                        int(vector_parameter[i])]
+            elif self.parameters[self.param_names[i]][0] == 'int':
+                dict_parameter[self.param_names[i]] = int(vector_parameter[i])
+            else:
+                dict_parameter[self.param_names[i]] = vector_parameter[i]
+
+        return dict_parameter
+
+    def _evaluate_params(self, X, y, params):
+        """
+        The score function to call in order to evaluate the quality
+        of the parameter params
+
+        Parameters
+        ----------
+        tested_parameter : dict, the parameter to test
+
+        Returns
+        -------
+        score : the mean of the CV score
```

Comment:
That's funny, I tried that when working on a related project and that's why there was the 'nugget' parameter for the GP. In my case it didn't work really well but that could be something to investigate. I also don't know if it's commonly done.

In any case I think we can propose two alternatives : either store only the mean or store all CV results. That could be useful for example if we'd like to maximize not the mean of the CV scores but something like mean-alpha*std to penalize a high standard deviation in CV results.